### PR TITLE
feat(earn): Add earn withdraw bottom sheet

### PR DESCRIPTION
**Pull request by finnian0826**
### Description

Add bottom sheet for different withdraw options, show if >=2 of those options exist for a user.

### Test plan
Unit tests added and updated

**Manual:**

Can partial withdraw:

https://github.com/user-attachments/assets/06db7b37-9a3a-462c-b93a-db7191b965d8

Can't partial withdraw:

https://github.com/user-attachments/assets/e675dd50-e3a2-4fab-b18e-8693c555fa42

### Related issues

- Fixes ACT-1385

### Backwards compatibility

Yes

### Network scalability

If a new NetworkId and/or Network are added in the future, the changes in this PR will:

- [X] Continue to work without code changes, OR trigger a compilation error (guaranteeing we find it when a new network is added)

### CODE REVIEW

File: src/earn/BeforeDepositBottomSheet.tsx [hunk 1]
```diff
@@ -25,14 +25,14 @@ import { Spacing } from 'src/styles/styles'
 import { TokenBalance } from 'src/tokens/slice'
 import { getTokenAnalyticsProps } from 'src/tokens/utils'
 
-function ActionCard({ action }: { action: BeforeDepositAction }) {
+export function ActionCard({ action }: { action: BeforeDepositAction | WithdrawAction }) {
```
Comment (satish-ravi) [inline src/earn/BeforeDepositBottomSheet.tsx:28]: Might be worth moving this into its own file. Maybe even share with other parts of the app (token details, etc), seems like it's something we use throughout the app and would be nice to change styles in a single place for mobile stack apps

File: src/earn/EarnPoolInfoScreen.test.tsx [hunk 2]
```diff
@@ -646,42 +658,240 @@ describe('EarnPoolInfoScreen', () => {
...
       </Provider>
     )
-    fireEvent.press(getByText('earnFlow.poolInfoScreen.withdraw'))
+    fireEvent.press(getByTestId('WithdrawButton'))
+    expect(getByTestId('Earn/WithdrawBottomSheet')).toBeVisible()
+    expect(getByTestId('Earn/ActionCard/Withdraw')).toBeTruthy()
+    expect(getByTestId('Earn/ActionCard/Claim')).toBeTruthy()
+    expect(getByTestId('Earn/ActionCard/Exit')).toBeTruthy()
+  })
+  it('tapping withdraw on WithdrawBottomSheet navigates to enter amount screen', () => {
```
Comment (satish-ravi) [inline src/earn/EarnPoolInfoScreen.test.tsx:701]: this file is getting too big, I've started breaking EarnPoolInfoScreen into multiple files on https://github.com/valora-inc/wallet/pull/6176, haven't broken out tests yet. But would it be possible to have all the bottom sheet specific tests in a `WithdrawBottomSheet.test.tsx`? and then `EarnPoolInfoScreen.test.tsx` can have a single test to ensure bottom sheet is opened.

File: src/icons/Claim.tsx
Comment (satish-ravi) [inline src/icons/Claim.tsx:1]: Can we name the icon Trophy? Would be good to keep it in sync with the name in the design system https://www.figma.com/design/erFfzHvSTm5g1sjK6jWyEH/Working-Design-System?node-id=2100-5070&node-type=frame&t=lvQvczdzHRWQBGKn-0

File: src/positions/types.ts [hunk 3]
```diff
@@ -25,6 +25,11 @@ export interface EarningItem {
   includedInPoolBalance?: boolean
 }
 
+export enum ClaimType {
```
Comment (satish-ravi) [inline src/positions/types.ts:28]: can this just be a string literal type instead of an enum?

File: src/analytics/Properties.tsx [hunk 4]
```diff
@@ -1639,6 +1639,7 @@ interface EarnEventsProperties {
     hasTokensOnSameNetwork: boolean
     hasTokensOnOtherNetworks: boolean
   }
+  [EarnEvents.earn_select_withdraw_type]: { type: 'partialWithdraw' | 'claim' | 'exit' }
```
Comment (satish-ravi) [inline src/analytics/Properties.tsx:1642]: nit: keep all the pool_info_tap ones together

File: src/analytics/docs.ts [hunk 5]
```diff
@@ -609,6 +609,7 @@ export const eventDocs: Record<AnalyticsEventType, string> = {
   [EarnEvents.earn_pool_info_tap_info_icon]: `When the user taps an info icon on the earn pool info screen`,
   [EarnEvents.earn_pool_info_tap_withdraw]: `When the user taps the withdraw button on the pool info screen`,
   [EarnEvents.earn_pool_info_tap_deposit]: `When the user taps the deposit button on the pool info screen`,
+  [EarnEvents.earn_select_withdraw_type]: `When a use selects a withdraw type on the withdraw bottom sheet`,
```
Comment (satish-ravi) [inline src/analytics/docs.ts:612]: same as above

File: src/earn/types.ts [hunk 6]
```diff
@@ -56,6 +56,12 @@ export enum BeforeDepositActionName {
   Swap = 'Swap',
 }
 
+export enum WithdrawActionName {
```
Comment (MuckT) [inline src/earn/types.ts:59]: Can this be a string literal type instead?

File: src/earn/EarnPoolInfoScreen.test.tsx (hunk 2 above)
Comment (finnian0826) [reply src/earn/EarnPoolInfoScreen.test.tsx:701]: @satish-ravi yeah I can do that, for having the bottom sheet tests in their own file, would I still render the pool info screen and then have it tap to open the bottom sheet, or is there a way to render just the bottoms sheet?

File: src/analytics/docs.ts [hunk 7]
```diff
@@ -610,6 +610,7 @@ export const eventDocs: Record<AnalyticsEventType, string> = {
   [EarnEvents.earn_pool_info_tap_withdraw]: `When the user taps the withdraw button on the pool info screen`,
   [EarnEvents.earn_pool_info_tap_deposit]: `When the user taps the deposit button on the pool info screen`,
   [EarnEvents.earn_pool_info_tap_safety_details]: `When the user taps the view more/less details on the safety card on the pool info screen`,
+  [EarnEvents.earn_select_withdraw_type]: `When a use selects a withdraw type on the withdraw bottom sheet`,
```
Comment (MuckT) [inline src/analytics/docs.ts:613]: nit
```suggestion
  [EarnEvents.earn_select_withdraw_type]: `When a user selects a withdrawal type on the withdraw bottom sheet`,
```

File: src/earn/EarnPoolInfoScreen.test.tsx (hunk 2 above)
Comment (finnian0826) [reply src/earn/EarnPoolInfoScreen.test.tsx]: moved to their own file in [5ceadc2](https://github.com/valora-inc/wallet/pull/6142/commits/5ceadc24807e5decc9ad1669180c0bd1115cd88f), did it similar to `EarnDepositBottomSheet.test.tsx` where I could just render the bottom sheet and seems to work as expected!

File: src/earn/types.ts [hunk 8]
```diff
@@ -64,4 +65,14 @@ export interface BeforeDepositAction {
...
+  name: WithdrawActionName
+  title: string
+  details: string
+  iconComponent: React.MemoExoticComponent<({ color }: { color: Colors }) => JSX.Element>
+  onPress: () => void
+}
+
 export type EarnEnterMode = 'deposit' | 'swap-deposit' | 'withdraw'
+
+export type EarnWithdrawType = 'partialWithdraw' | 'claim' | 'exit'
```
Comment (MuckT) [inline src/earn/types.ts:78]: `EarnWithdrawType` & `WithdrawActionName` look similar is there a reason not to combine them?

Comment (finnian0826) [reply src/earn/types.ts]: good catch! updated in [3f4abb2](https://github.com/valora-inc/wallet/pull/6142/commits/3f4abb259406ae5cfedcf8b8c65778c428b5baed)

File: locales/base/translation.json [hunk 9]
```diff
@@ -2713,6 +2713,20 @@
...
       "withdraw": "Withdraw",
+      "withdrawBottomSheet": {
+        "title": "Select withdraw type",
+        "withdrawAndClaim": "Withdraw and Claim",
+        "withdrawDescription": "Withdraw a specific amount of any tokens you have deposited in the pool",
+        "withdrawAndClaimDescription": "Withdraw a specific amount of any tokens you have deposited in the pool and claim your rewards automatically",
+        "claimRewards": "Claim Rewards",
+        "claimRewardsDescription": "Claim your rewards. Partial amounts of the reward cannot be claimed",
+        "claimEarnings": "Claim Earnings",
+        "claimEarningsDescription": "Claim the earnings on your deposit. Partial amounts cannot be claimed ",
```
Comment (MuckT) [inline locales/base/translation.json:2724]:
```suggestion
        "claimEarningsDescription": "Claim the earnings on your deposit. Partial amounts cannot be claimed",
```

**Review by MuckT — DISMISSED** [review]

File: src/utils/ActionCard.tsx
Comment (satish-ravi) [inline src/utils/ActionCard.tsx:1]: we generally don't put components in utils, you can leave this at src/earn. I believe eventually this should go to src/components and reused across the other bottom sheets. Can you make an issue for it?

File: src/earn/poolInfoScreen/WithdrawBottomSheet.tsx [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    <BottomSheet
+      forwardedRef={forwardedRef}
+      title={t('earnFlow.poolInfoScreen.withdrawBottomSheet.title')}
+      titleStyle={styles.bottomSheetTitle}
+      testId={'Earn/WithdrawBottomSheet'}
+    >
+      <View style={styles.actionsContainer}>
+        {canPartialWithdraw && (
+          <PartialWithdrawAction forwardedRef={forwardedRef} pool={pool} hasRewards={canClaim} />
+        )}
```
Comment (satish-ravi) [inline src/earn/poolInfoScreen/WithdrawBottomSheet.tsx:136]: is the gate check needed here? this bottom sheet will never show up if the gate is disabled per changes in EarnPoolInfoScreen right?

File: src/earn/poolInfoScreen/WithdrawBottomSheet.test.tsx [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+        isCashInEligible: true,
+      },
+    },
+  },
+  positions: {
+    positions: [...mockPositions, ...mockRewardsPositions],
+  },
+})
+
+it('tapping withdraw on WithdrawBottomSheet navigates to enter amount screen', () => {
```
Comment (satish-ravi) [inline src/earn/poolInfoScreen/WithdrawBottomSheet.test.tsx:41]: wrap these in a describe

File: src/earn/poolInfoScreen/WithdrawBottomSheet.test.tsx [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+        canClaim={true}
+      />
+    </Provider>
+  )
+  expect(getByTestId('Earn/ActionCard/PartialWithdraw')).toBeTruthy()
+  fireEvent.press(getByTestId('Earn/ActionCard/PartialWithdraw'))
+  expect(AppAnalytics.track).toHaveBeenCalledWith(EarnEvents.earn_select_withdraw_type, {
+    type: 'PartialWithdraw',
+  })
+  expect(navigate).toHaveBeenCalledWith(Screens.EarnEnterAmount, {
```
Comment (satish-ravi) [inline src/earn/poolInfoScreen/WithdrawBottomSheet.test.tsx:59]: needs a jest.clearAllMocks in a `beforeEach` to ensure clean assertions

File: src/earn/poolInfoScreen/EarnPoolInfoScreen.tsx [hunk 13]
```diff
@@ -209,10 +218,10 @@ export default function EarnPoolInfoScreen({ route, navigation }: Props) {
     const partialWithdrawalsEnabled = getFeatureGate(
       StatsigFeatureGates.ALLOW_EARN_PARTIAL_WITHDRAWAL
     )
-    if (partialWithdrawalsEnabled) {
-      navigate(Screens.EarnEnterAmount, { pool, mode: 'withdraw' })
+    if (hasRewards || partialWithdrawalsEnabled) {
+      withdrawBottomSheetRef.current?.snapToIndex(0)
     } else {
-      navigate(Screens.EarnCollectScreen, { pool })
+      navigate(Screens.EarnCollectScreen, { pool }) // TODO (ACT-1389): Confirmation screen for Claim & Withdraw
```
Comment (satish-ravi) [inline src/earn/poolInfoScreen/EarnPoolInfoScreen.tsx:224]: question on the todo here. should we keep the gate off flow the current one? i.e., all changes for partial withdrawals  / new screens go behind the gate, cc @MuckT

File: src/earn/poolInfoScreen/EarnPoolInfoScreen.tsx [hunk 14]
```diff
@@ -198,6 +202,11 @@ export default function EarnPoolInfoScreen({ route, navigation }: Props) {
     exchanges,
   } = useDepositEntrypointInfo({ allTokens, pool })
 
+  const rewardsPositions = useSelector(positionsWithBalanceSelector).filter((position) =>
+    pool.dataProps.rewardsPositionIds?.includes(position.positionId)
+  )
+  const hasRewards = useMemo(() => rewardsPositions.length > 0, [rewardsPositions])
```
Comment (satish-ravi) [inline src/earn/poolInfoScreen/EarnPoolInfoScreen.tsx:208]: can the filter also go inside the useMemo?

File: src/analytics/Properties.tsx [hunk 15]
```diff
@@ -1642,6 +1647,7 @@ interface EarnEventsProperties {
   [EarnEvents.earn_pool_info_tap_safety_details]: EarnCommonProperties & {
     action: 'expand' | 'collapse'
   }
+  [EarnEvents.earn_select_withdraw_type]: { type: WithdrawActionName }
```
Comment (satish-ravi) [inline src/analytics/Properties.tsx:1650]: can we include EarnCommonProperties here?

File: src/earn/poolInfoScreen/EarnPoolInfoScreen.tsx (hunk 13 above)
Comment (MuckT) [reply src/earn/poolInfoScreen/EarnPoolInfoScreen.tsx:224]: We should fix this so the old flow is served by default until the feature gate is enabled.

File: src/earn/poolInfoScreen/WithdrawBottomSheet.tsx (hunk 10 above)
Comment (finnian0826) [reply src/earn/poolInfoScreen/WithdrawBottomSheet.tsx:136]: I think it is needed, in EarnPoolInfoScreen, we show the bottom sheet if `(hasRewards || partialWithdrawalsEnabled)`, so if the user has rewards but the gate is disabled, then the bottom sheet will still be shown with the options to claim rewards or exit.

File: src/earn/poolInfoScreen/EarnPoolInfoScreen.tsx (hunk 13 above)
Comment (finnian0826) [reply src/earn/poolInfoScreen/EarnPoolInfoScreen.tsx:228]: @satish-ravi @MuckT are you suggesting that we use `ALLOW_EARN_PARTIAL_WITHDRAWAL` as the gate to decide whether the new flow is shown, or adding an additional `SHOW_EARN_WITHDRAW_OPTIONS` gate on top of the existing one?

**Review by MuckT — APPROVED** [review]

**Pull request merged**